### ISSUE TITLE
Provide a non-indy version of the library

### ISSUE THREAD

**Issue by marcos-carceles** [elasticsearch/elasticsearch-groovy#20]
Hi,

first of all thank you very much for the great work on building a groovy version of the library. I found it really useful and simplifies a lot using elasticsearch for groovy users.

It would be really useful for many teams to be able to use a non indy version of the library. One of the reasons for needing this would be if they would be using a framework (like Grails) which already ships with a non indy version of groovy-all. On these scenarios it would be quite complex (if possible at all) to switch the framework to the indy version of groovy and unfortunately both versions can't coexist on the same classpath (as seen here: http://groovy.codehaus.org/InvokeDynamic+support). Another (less likely) reason would be if they were using a non-indy compatible jvm (< java 7)

I'm currently working on both a Grails application and a Grails plugin for using elasticsearch (https://github.com/noamt/elasticsearch-grails-plugin). I tried to replace the elasticsearch client with the elasticsearch-groovy client on our plugin, but I found problems due to the indy version of the library.

Many thanks in advance.

**Comment by pickypg** [comment]
@marcos-carceles Thanks for using the client. I'm glad that it is helping you.

Considering the inability for Grails 2.x to use unbundled versions of Groovy, I agree that releasing a non-indy jar makes a lot of sense. I will try to release a "no-indy" variant that can be used in such cases with the upcoming release of 1.4.3 (soon).

> Another (less likely) reason would be if they were using a non-indy compatible jvm (< java 7)

Elasticsearch requires Java 7. As such, this is a minimum requirement of the Groovy client as well because it depends on the Elasticsearch jar and that won't change even after I add a no-indy jar.

Fortunately, it looks like [Grails 3 will support the ability to change the bundled Groovy version](https://jira.grails.org/browse/GRAILS-8339), so I will probably provide a no-indy jar _at least_ until that becomes mainstream. Hopefully [when Java 7 goes EOL in April](http://www.oracle.com/technetwork/java/eol-135779.html#Java6-end-public-updates), usage of indy will become more common within the Groovy community by _at least_ no longer supporting Java 6 installations.

**Comment by marcos-carceles** [comment]
Hi,

after having a play with it, I noticed that once disabled invokedynamic, randomized is making the tests to fail as it does not accept Groovy injected fields (more specifically `callSiteArray`). I noticed that `GroovyTestSanitizer` already deals with a similar situation and I have included  a workaround to accept these fields. Apart from that, from what I've seen so far everything seems to work alright.

I haven't created a Pull Request yet, because I assumed you'd want to have a new branch for the non-indy version.

Cheers!

**Comment by marcos-carceles** [comment]
Hi @pickypg,

is there any way I can contribute to help releasing this?

**Comment by pickypg** [comment]
@marcos-carceles I've been distracted for the past week. I am happy to accept pull requests, but I really want to make sure that the noindy jar is created in parallel to the indy (normal) jar without having separate branches.

This will help to guarantee that code is always merged over and that unit tests work against both. I am hopeful that I can get the build working tonight and make the 1.4.3 release this week.

**Comment by pickypg** [comment]
@marcos-carceles I have pushed the change that makes this work, but I am testing with some Grails packages (namely yours) before proceeding.

The Maven/Gradle accessible name is

```gradle
runtime 'org.elasticsearch:elasticsearch-groovy:1.4.3-SNAPSHOT:grails'
```

```xml
<dependency>
  <groupId>org.elasticsearch</groupId>
  <artifactId>elasticsearch-groovy</artifactId>
  <version>1.4.3-SNAPSHOT</version>
  <classifier>grails</classifier>
  <scope>runtime</scope>
</dependency>
```

I am having trouble getting Grails to recognize the classifier without specifying it as:

```groovy
mavenRepo "https://oss.sonatype.org/content/repositories/snapshots/"

// ...

runtime group: 'org.elasticsearch', name: 'elasticsearch-groovy', version: '1.4.3-SNAPSHOT', classifier: 'grails'
```

Perhaps it needs to add, but it seems to work without excluding groovy:

```groovy
runtime group: 'org.elasticsearch', name: 'elasticsearch-groovy', version: '1.4.3-SNAPSHOT', classifier: 'grails', {
  excludes "groovy-all"
}
```

Could you test this and verify that it is working with your project? I cloned your repo and ran the tests (`./grailsw test-app`). Assuming that it does work, then I should be able to push out the release tonight or tomorrow.

**Comment by pickypg** [comment]
I created quick test project with Grails using just a domain and controller. I added some code to connect to the cluster (unrelated to the actual code purpose of the domain/controller code) and it was invoked correctly.

I also modified your `ClientNodeFactoryBean` class to do a search at the end of it and that worked using the Groovy enhancements.

This jar will require Grails 2.4.4 in order to be able to benefit from the fix in [GRAILS-10652](https://jira.grails.org/browse/GRAILS-10652) (specifically https://github.com/grails/grails-core/commit/a817523233e98aab32e593d00b238e7d54f02bd9, which is in both 2.4.4 and 3.0.0m1). The issue fixes using Groovy Extension Modules, which the Elasticsearch Groovy client uses extensively. Earlier versions will require adding the issue-attached code to customize builds to force the behavior.
